kernel/src/syscall.rs

```diff
@@ -522,111 +552,197 @@ impl SyscallReturn {
     /// Encode the system call return value into 4 registers, following the
     /// encoding specified in TRD104. Architectures which do not follow TRD104
     /// are free to define their own encoding.
+    /// TODO: deprecate in favour of the more general one
     pub fn encode_syscall_return(&self, a0: &mut u32, a1: &mut u32, a2: &mut u32, a3: &mut u32) {
+        if core::mem::size_of::<MetaPtr>() == core::mem::size_of::<u32>() {
+            // SAFETY: if the two unsized integers are the same size references to them
+            // can be safely transmuted.
+            // Ugly coercion could be avoided by first copying to the stack, then assigning with
+            // "as" in order to satisfy the compiler. But I expect this function will disappear
+            // in favour of just using the usize one.
+            unsafe {
+                let a0 = &mut *(core::ptr::from_mut(a0) as *mut MetaPtr);
+                let a1 = &mut *(core::ptr::from_mut(a1) as *mut MetaPtr);
+                let a2 = &mut *(core::ptr::from_mut(a2) as *mut MetaPtr);
+                let a3 = &mut *(core::ptr::from_mut(a3) as *mut MetaPtr);
+                self.encode_syscall_return_mptr(a0, a1, a2, a3);
+            }
+        } else {
+            panic!("encode_syscall_return used on a 64-bit platform or CHERI platform")
+        }
+    }
+
+    /// The obvious extension of TRD104 that works for 32-bit and 64-bit platforms.
+    /// This makes no changes to TRD104 on 32-bit platforms.
+    /// On 64-bit platforms, both 64-bit and usize values are passed as a single register,
+    /// shifting down register number if that means fewer registers are needed needed.
+    /// For usize, this is the obvious choice.
+    /// For explicitly 64-bit arguments, this would require rewriting prototypes for userspace
+    /// functions between 32 and 64 bit platforms.
+    /// However, no driver currently USES any 64-bit values.
+    /// Any new ones on 64-bit platforms would likely prefer just passing the value.
+    /// If they would not, I suspect many really want usize anyway (and that is what is used for
+    /// the syscalls handled directly by the kernel). Maybe they should be written in terms of that,
+    /// and some helpful aliases created for FailureUSIZE etc.
+    /// I think packing two 32-bit values into 64-bits would gain nothing and pollute a whole lot
+    /// of user code.
+    /// I have not considered usize other than 4 and 8 bytes.
+    /// Also handles the CHERI extension as follows:
+    /// the high part of any MetaPtr register is zero'd if any non capability-sized arguments are
+    /// passed.
+    /// SuccessPtr is as passed the full MetaPtr register.
+    /// Pointers from allow'd buffers have minimal bounds attached that cover their length,
+    /// and the same permissions that were checked at the syscall boundary.
+    pub fn encode_syscall_return_mptr(
```

Comment:
Can we move this to the CHERI architecture, where it's actually used? I'd prefer implementations only used on one platform to be shipped in that platform support code, instead of generic kernel code.

Reply:
This one in intended to work for 32/64 bit CHERI / non-CHERI and so I think belongs in kernel.